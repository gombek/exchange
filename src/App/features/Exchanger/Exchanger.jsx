import React, { Component } from 'react';
import CurrencyPanel from '../../components/CurrencyPanel';
import CurrentRate from '../../components/CurrentRate';
import Button from '../../atomics/Button';

import './Exchanger.scss';

const onValueChange = (handler, rate) => value => {
  if (typeof handler !== 'function') return;
  handler(value, rate);
};

const onSubmitHandler = handler => ({from, to, rate, amount}) => () => {
  if (typeof handler !== 'function') return;
  handler({from, to, rate, amount});
};

class Exchanger extends Component {
  componentDidMount() {
    this.props.startLiveRates();
  }

  componentWillUnmount() {
    this.props.stopLiveRates();
  }

  render() {
    let {
      rate,
      currencies,
      onSubmit,
      from,
      fromValue,
      fromBalance,
      onFromValueChange,
      onFromCurrencyChange,
      to,
      toValue,
      toBalance,
      onToValueChange,
      onToCurrencyChange,
    } = this.props;

    return (
      <div className="exchanger">
        <CurrencyPanel
          value={fromValue}
          balance={fromBalance}
          activeCurrency={from}
          currencies={
            /* Locked to USD due to the API subscription limitation.
               For different currencies, the exchange rates
               wont be fetch until the subscription upgrade.
               Change to `currencies` to unlock. */
            [from]
          }
          onValueChange={onValueChange(onFromValueChange, rate)}
          onCurrencyChange={onFromCurrencyChange}
        />
        <div className="exchanger__rate">
          <CurrentRate
            currency={from}
            exchangeTo={to}
            exchangeRate={rate}
          />
        </div>
        <CurrencyPanel
          className="exchanger__to"
          value={toValue}
          balance={toBalance}
          activeCurrency={to}
          currencies={currencies}
          onValueChange={onValueChange(onToValueChange, rate)}
          onCurrencyChange={onToCurrencyChange}
        />
        <div className="exchanger__actions">
          <Button
            onClick={onSubmitHandler(onSubmit)({ from, to, rate, amount: fromValue })}
          >
            Exchange
          </Button>
        </div>
      </div>
    );
  }
}

export default Exchanger;