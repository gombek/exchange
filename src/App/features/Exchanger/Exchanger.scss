@import "../../theme.scss";
.exchanger {
    display: flex;
    flex-direction: column;
    flex: 1;

    &__from {
        background-color: $grey-900-color;
    }

    &__to {
        background-color: $grey-800-color;
    }

    &__rate {
        box-sizing: border-box;
        padding: 5px 10px;
        display: flex;
        justify-content: center;
        background: linear-gradient(to bottom, $grey-900-color 50%, $grey-800-color 50%) bottom left repeat-x;
    }

    &__actions {
        background-color: $grey-800-color;
        box-sizing: border-box;
        padding: 10px 25px 15px;
        display: flex;
        flex-direction: column;
    }
}