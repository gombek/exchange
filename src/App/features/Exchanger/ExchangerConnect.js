import { connect } from 'react-redux';
import Exchanger from './Exchanger';
import { wallet, currencies, walletRates, exchanger } from '../../../modules';


const mapStateToProps = (state) => {
  const from = exchanger.selectors.getFrom(state);
  const to = exchanger.selectors.getTo(state);

  return {
    rate: walletRates.selectors.getExchange(state, from, to),
    currencies: currencies.selectors.get(state),
    from,
    fromValue: exchanger.selectors.getAmount(state),
    fromBalance: wallet.selectors.getBalance(state, from),
    to,
    toValue: exchanger.selectors.getOutcome(state),
    toBalance: wallet.selectors.getBalance(state, to),
  }
};

const mapDispatchToProps = dispatch => ({
  onFromValueChange(value, rate) {
    dispatch(exchanger.actions.setAmount({value, rate}));
  },
  onFromCurrencyChange(currency) {
    dispatch(exchanger.actions.setFrom(currency));
  },
  onToValueChange(value, rate) {
    dispatch(exchanger.actions.setOutcome({value, rate}));
  },
  onToCurrencyChange(currency) {
    dispatch(exchanger.actions.setTo(currency));
  },
  onSubmit({from, to, rate, amount}) {
    dispatch(wallet.actions.exchange({from, to, rate, amount}));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(Exchanger)