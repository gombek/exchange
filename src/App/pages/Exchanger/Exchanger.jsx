import React from 'react';
import CurrencySlide from '../../features/CurrencySlide';

const Exchanger = () => (
  <CurrencySlide currency="USD" exchangeTo="PLN" />
);

export default Exchanger;