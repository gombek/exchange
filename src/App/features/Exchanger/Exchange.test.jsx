import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Exchanger from './Exchanger';

describe('Exchanger feature', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Exchanger/>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders without crashing', () => {
    const nodes = renderer
      .create(
        <Exchanger
          rate={ 2 }
          currencies={ ['USD', 'EUR'] }
          from={ 'USD' }
          fromValue={ 100 }
          fromBalance={ 2000 }
          to={ 'EUR' }
          toValue={ 200 }
          toBalance={ 0 }
        />
      )
      .toJSON();
    expect(nodes).toMatchSnapshot();
  });
});
