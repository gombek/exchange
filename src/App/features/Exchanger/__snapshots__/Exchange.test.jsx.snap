// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchanger feature renders without crashing 1`] = `
<div
  className="exchanger"
>
  <div
    className="currency-panel"
  >
    <div
      className="currency-select"
    >
      <select
        className="currency-select__native"
        onChange={[Function]}
        value="USD"
      >
        <option
          value="USD"
        >
          USD
        </option>
      </select>
      <span
        className="currency-select__label"
      >
        USD
      </span>
    </div>
    <div
      className="input"
    >
      <input
        className="input__field"
        min={0}
        onChange={[Function]}
        placeholder={0}
        step=".01"
        type="number"
        value={100}
      />
    </div>
    <div
      className="currency-panel__balance"
    >
      You have: 
      <span>
        <abbr>
          $
           
        </abbr>
        2000.00
      </span>
    </div>
  </div>
  <div
    className="exchanger__rate"
  >
    <span
      className="current-rate"
    >
      <span
        className="chart-icon"
      >
        <svg
          id="Capa_1"
          version="1.1"
          viewBox="0 0 612.002 460.000"
          x="0px"
          xmlns="http://www.w3.org/2000/svg"
          y="0px"
        >
          <g>
            <g>
              <path
                d="M108.485,376.479c4.436,0,8.927-1.25,12.921-3.85l133.784-87.186l116.218,51.979c3.108,1.39,6.404,2.067,9.678,2.067 c5.84,0,11.604-2.158,16.069-6.277l112.802-103.978v18.584c0,13.094,10.613,23.709,23.707,23.709 c13.094,0,23.707-10.615,23.707-23.709V175.56c0-13.094-10.613-23.709-23.707-23.709h-72.26 c-13.094,0-23.707,10.615-23.707,23.709s10.613,23.707,23.707,23.707h11.102l-96.003,88.493l-113.755-50.877 c-3.092-1.383-6.392-2.065-9.676-2.065c-4.529,0-9.035,1.298-12.946,3.846L95.518,332.904c-10.97,7.148-14.069,21.834-6.92,32.809 C93.142,372.684,100.735,376.479,108.485,376.479z"
              />
            </g>
          </g>
        </svg>
      </span>
      <span>
        <abbr>
          $
           
        </abbr>
        1
      </span>
      <span>
         = 
      </span>
      <span>
        <abbr>
          €
           
        </abbr>
        2.00000
      </span>
    </span>
  </div>
  <div
    className="currency-panel exchanger__to"
  >
    <div
      className="currency-select"
    >
      <select
        className="currency-select__native"
        onChange={[Function]}
        value="EUR"
      >
        <option
          value="USD"
        >
          USD
        </option>
        <option
          value="EUR"
        >
          EUR
        </option>
      </select>
      <span
        className="currency-select__label"
      >
        EUR
      </span>
    </div>
    <div
      className="input"
    >
      <input
        className="input__field"
        min={0}
        onChange={[Function]}
        placeholder={0}
        step=".01"
        type="number"
        value={200}
      />
    </div>
    <div
      className="currency-panel__balance"
    >
      You have: 
      <span>
        <abbr>
          €
           
        </abbr>
        0.00
      </span>
    </div>
  </div>
  <div
    className="exchanger__actions"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      <span
        className="button__label"
      >
        Exchange
      </span>
    </button>
  </div>
</div>
`;
